larpandoracontent/LArTwoDReco/LArClusterAssociation/MvaLowEClusterMergingAlgorithm.cc

```diff
@@ -0,0 +1,623 @@
+template <typename T>
+StatusCode MvaLowEClusterMergingAlgorithm<T>::EdgeHitComparer(const pandora::ClusterList *const pClusterList, const std::string &listName ) const
+{
+    std::map<const Cluster*, const MCParticle*> clusterToMCParticleMap;
+    std::map<const Cluster*, bool> clusterIsUsed;
+    std::map<const Cluster*, ClusterVector> clustersToMerge;
+    int sigBac{0};
+    int countHits{0};
+
+    for (const Cluster *const pCluster : *pClusterList)
+    {
+	countHits = countHits + (pCluster->GetNCaloHits());
+    }
+
+    for (auto iter = pClusterList->begin(); iter != pClusterList->end(); ++iter)
+    {
+       const Cluster *const cluster(*iter);
+       const MCParticle *clusterMC(this->GetMCForCluster(cluster, clusterToMCParticleMap));
+       const int nClusters(pClusterList->size());
+       
+       if (!this->IsValidToUse(cluster, clusterIsUsed))
+            continue;
+
+	CaloHitList clusterEdgeHits;
+	CartesianVector centroid11(0.f, 0.f, 0.f);
+        CartesianVector centroid12(0.f, 0.f, 0.f);
+        try {
+       	centroid11 = cluster->GetCentroid(cluster->GetInnerPseudoLayer());
+        centroid12 = cluster->GetCentroid(cluster->GetOuterPseudoLayer());
+        }
+        catch (...) {
+        CaloHitList hits;
+        cluster->GetOrderedCaloHitList().FillCaloHitList(hits);
+        centroid11 = hits.front()->GetPositionVector();
+        centroid12 = hits.front()->GetPositionVector();
+        }
+
+        const CartesianVector centroid1((centroid11 + centroid12) * 0.5f);
+	
+	this->EdgeHitFinder(cluster, clusterEdgeHits); 
+
+        const unsigned int clusterNHits{cluster->GetNCaloHits()};
+        const unsigned long int clusterNEdgeHits{clusterEdgeHits.size()};
+
+	for (auto iter2 = std::next(iter) ; iter2 != pClusterList->end() ; ++iter2)
+	{
+	    const Cluster *const otherCluster(*iter2);
+	    const MCParticle *otherClusterMC(this->GetMCForCluster(otherCluster, clusterToMCParticleMap));
+            sigBac = 0;
+
+            if (clusterMC == otherClusterMC)
+            {
+                   sigBac = 1;
+            }
+
+	    CaloHitList otherClusterEdgeHits;
+	    CartesianVector centroid21(0.f, 0.f, 0.f);
+	    CartesianVector centroid22(0.f, 0.f, 0.f);
+	    try {
+	    centroid21 = otherCluster->GetCentroid(otherCluster->GetInnerPseudoLayer());
+            centroid22 = otherCluster->GetCentroid(otherCluster->GetOuterPseudoLayer());
+	    }
+	    catch (...) {
+	    CaloHitList hits;
+            cluster->GetOrderedCaloHitList().FillCaloHitList(hits);
+            centroid21 = hits.front()->GetPositionVector();
+            centroid22 = hits.front()->GetPositionVector();
+	    }
+	    const CartesianVector centroid2((centroid21 + centroid22) * 0.5f);
+	    const float centroidSeparation{std::sqrt(centroid1.GetDistanceSquared(centroid2))};
+	    const CartesianVector centroidVector(centroid2.GetX() - centroid1.GetX(), 0.f, centroid2.GetZ() - centroid1.GetZ());
```

Comment:
Doesn't Pandora already provide functionality to calculate vector addition and subtraction?